From d3c08e9ef800f3eb24801efcde08434e353ff204 Mon Sep 17 00:00:00 2001
From: skc <chaitanya.sakinam@nxp.com>
Date: Tue, 20 Feb 2018 09:59:29 +0530
Subject: [PATCH] PPSBR-1995 PPSBR-1996 PPSBR-977 PPSBR-978 IPSEC support on
 PPTP/L2TP WAN types

Support for IPSec on PPTP/L2TP WAN connection types

Signed-off-by: skc <chaitanya.sakinam@nxp.com>
---
 net/key/af_key.c | 27 ++++++++++++++++++++++++++-
 1 file changed, 26 insertions(+), 1 deletion(-)

diff --git a/net/key/af_key.c b/net/key/af_key.c
index 02d1593..8897085 100644
--- a/net/key/af_key.c
+++ b/net/key/af_key.c
@@ -4180,6 +4180,7 @@ static struct sk_buff * ipsec_xfrm2nlkey (struct net *net, struct xfrm_state *x,
 	struct nlmsghdr *nlh = NULL;
 	gfp_t allocation = GFP_ATOMIC; //This may called from atomic context
 	unsigned char tunnel, keys, natt, state, lifetime;
+	struct rtable *rt;
 
 	/* supported SA informations */
 	keys = 1; state = 1; tunnel = 1; lifetime = 1; natt = 1; 
@@ -4200,7 +4201,31 @@ static struct sk_buff * ipsec_xfrm2nlkey (struct net *net, struct xfrm_state *x,
 					printk(KERN_ERR "ipsec_xfrm2nlkey: protocol %d not supported\n", x->id.proto);
 					*msg_id = NLKEY_NULL_MSG;
 					goto exit;
-				}	
+
+				}
+				rt = ip_route_output(net, x->id.daddr.a4, 0, 0, 0);
+				if (!IS_ERR(rt)) {
+					if ((!strncmp(rt->dst.dev->name,"l2tp", 4)|| !strncmp(rt->dst.dev->name,"pptp", 4))) {
+						//printk(KERN_INFO "Dst is l2tp/pptp iface %s, no sa offload\n",\
+							rt->dst.dev->name);
+						*msg_id = NLKEY_NULL_MSG; /* next message */
+							ip_rt_put(rt);
+						goto exit;
+						}
+					ip_rt_put(rt);
+				}
+				rt = ip_route_output(net, x->props.saddr.a4, 0, 0, 0);
+				if (!IS_ERR(rt)) {
+					if ((!strncmp(rt->dst.dev->name,"l2tp", 4)|| !strncmp(rt->dst.dev->name,"pptp", 4))) {
+						//printk(KERN_INFO "Src is l2tp/pptp if %s, no sa offload\n",\
+							rt->dst.dev->name);
+						*msg_id = NLKEY_NULL_MSG; /* next message */
+						ip_rt_put(rt);
+						goto exit;
+					}
+					ip_rt_put(rt);
+				}
+
 				memset(&sa_create_msg, 0, sizeof(struct nlkey_sa_create));	
 
 				/* SA global handler */
-- 
1.9.1

